### DISCUSSION TITLE
Language server is not re-installed properly when using goenv

### DISCUSSION THREAD

**Discussion by mhughdo** [zed-industries/zed#15684] [Feature Requests]
### Issue description

The [get_language_server_command](https://github.com/zed-industries/zed/blob/b88b9dcdd17e6c2c587f276cae8320cc888d027d/crates/language/src/language.rs#L279) function in LspAdapter returns the language server binary if it is already installed or cached. If not, it will download the language server.

When using goenv, a gopls binary file is installed at ~/.goenv/shims/gopls. This gopls binary will only work properly if there is a gopls binary installed for a specific Go version at ~/go/<go-version>/bin.

The [start_language_server](https://github.com/zed-industries/zed/blob/b88b9dcdd17e6c2c587f276cae8320cc888d027d/crates/project/src/project.rs#L3037) function in project.rs will find and start the gopls binary at ~/.goenv/shims/gopls. When gopls returns errors, it will attempt to reinstall the language server. However, the problem lies in the fact that get_language_server_command is called again, and it returns the same gopls binary that caused the errors, as it is in the path. As a result, Zed does not attempt to install a new version. When adapter.reinstall_attempt_count reaches its maximum value, the error `Hit {max} reinstallation attempts for {server_name:?}` is logged and no working gopls will be installed.
![telegram-cloud-photo-size-5-6154394342681330804-y](https://github.com/user-attachments/assets/5d44c532-5f1b-497d-8ddd-b95d71caee76)

The problem occurs each time a new Go version is installed and could be fixed by manually installing the gopls but I think It'd be great if Zed could do it automatically.

### Proposed solution

- Add some kind of flag (e.g is_reinstall: bool) to the get_language_server_command function which will be used to ignore/delete the gopls in the PATH and install a new global gopls binary.

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
